operator_cxx/contrib/quantization_int8-inl.h

```diff
@@ -0,0 +1,505 @@
+                  mshadow::Copy(aux, max_val, s);
+                }
+                init = false;
+            }
+            else {
+              DType old_aux = DType(0.0f);
+              Tensor<cpu, 1, DType> old_aux_tensor(&old_aux, Shape1(1), s_cpu);
+              mshadow::Copy(old_aux_tensor, aux, s);
+              mshadow::Copy(threshold_tensor, max_val, s);
+              old_aux = param_.ema_decay * old_aux + (1 - param_.ema_decay) * threshold;
+              mshadow::Copy(aux, old_aux_tensor, s);
+            }
+          }
+          // DType quant_unit = aux[0] / QUANT_LEVEL;
+          
+          mshadow::Copy(threshold_tensor, aux, s);
+          DType quant_unit = threshold / QUANT_LEVEL;
+          const ScalarExp<DType> quant_unit_expr(quant_unit);
+          const ScalarExp<DType> threshold_expr(threshold);
+          Assign(out, req[Quantization_int8_enum::kOut], 
+                 F<mshadow_op::round>(F<mshadow_op::clip>(data, threshold_expr) /quant_unit_expr) * quant_unit_expr);
+        }
+      }
+      else {
+        LOG(FATAL) << "quantization int8 only support minmax mode currently";
+      }
+    }
+  }
+  
+
+  virtual void Backward(const OpContext & ctx,
+                        const std::vector<TBlob> &out_grad,
+                        const std::vector<TBlob> &in_data,
+                        const std::vector<TBlob> &out_data,
+                        const std::vector<OpReqType> &req,
+                        const std::vector<TBlob> &in_grad,
+                        const std::vector<TBlob> &aux_states) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    using namespace mxnet_op;
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 4, DType> data_grad;
+    Tensor<xpu, 4, DType> out_data_grad;
+    Tensor<xpu, 4, DType> data;
+    Tensor<xpu, 4, DType> out;
+    if (out_grad[Quantization_int8_enum::kOut].ndim() == 2) {
+      Shape<4> dshape = Shape4(out_grad[Quantization_int8_enum::kOut].shape_[0],
+                               out_grad[Quantization_int8_enum::kOut].shape_[1], 1, 1);
+      data = in_data[Quantization_int8_enum::kData].get_with_shape<xpu, 4, DType>(dshape, s);
+      out = out_data[Quantization_int8_enum::kOut].get_with_shape<xpu, 4, DType>(dshape, s);
+
+      out_data_grad = out_grad[Quantization_int8_enum::kOut].get_with_shape<xpu, 4, DType>(dshape, s);
+      data_grad = in_grad[Quantization_int8_enum::kData].get_with_shape<xpu, 4, DType>(dshape, s);
+    } else {
+      data = in_data[Quantization_int8_enum::kData].get<xpu, 4, DType>(s);
+      out = out_data[Quantization_int8_enum::kOut].get<xpu, 4, DType>(s);
+
+      out_data_grad = out_grad[Quantization_int8_enum::kOut].get<xpu, 4, DType>(s);
+      data_grad = in_grad[Quantization_int8_enum::kData].get<xpu, 4, DType>(s);
+    }
+    // Assign(data_grad, req[Quantization_int8_enum::kOut], out_data_grad);
+    if (param_.grad_mode == std::string("ste") || param_.is_weight) {
```

Comment:
What is ste

Reply:
Straight-Through-Estimator, pass the `out gradients` to `data gradients` directly.